### PR TITLE
Improve changes validator api

### PR DESCRIPTION
**Pull request by dimagorelik**
Improve changes validator API

### CODE REVIEW

File: main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import com.kenshoo.pl.entity.EntityType;
+import com.kenshoo.pl.entity.internal.EntityTypeReflectionUtil;
+import com.kenshoo.pl.entity.internal.validators.*;
+import com.kenshoo.pl.entity.spi.*;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class OptimizedChangesValidator<E extends EntityType<E>> implements ChangesValidator<E> {
```
Comment (dimagorelik) [inline main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @galkoren @effiban 
Do you like this name? Do you have another suggestion?

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: I don't understand why you call it "optimized" - what is the optimization ?

File: main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    @Override
+    public Stream<EntityField<?, ?>> requiredFields(Collection<? extends EntityField<E, ?>> fieldsToUpdate, ChangeOperation changeOperation) {
+        return findValidatorsTriggeredByFields(fieldsToUpdate, changeOperation)
+                .flatMap(ChangeValidatorAdapter::fieldsToFetch);
+    }
+
+    public void register(ChangeValidatorAdapter<E> validatorAdapter) {
+        triggeredChangeValidators.add(validatorAdapter);
+    }
```
Comment (effiban) [inline main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:46]: Shouldn't this be inside the builder as well?

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java]: @effiban 
Fixed

File: main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java (hunk 1 above)
Comment (galkoren) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: Name is tricky... this class is yet another adaptor, an adaptor of all other adaptors to become a single Validator.
So it is a composite. But we already have many other composites. We need to discuss it.

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: As I understand the difference from the previous impl is that we now have a fluent way to add validators - right?
If so we can call it FluentCompositeChangeValidator

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @effiban 
Please, pay attention that this validator implements the general ChangesValidator interface

The validator does the number of optimization:
1. Calculate fields to fetch according to updated fields
2. Apply only validators that triggered by updated fields

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: Ok, but you didn't actually change the logic compare to the old name, right? You just added the builder?

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: I don't think that it is more or less "optimized" than anything else - all the validators should calculate fields to fetch and be applied only when needed.
It is composite like Gal said, and it also uses adapters.
So maybe we could call it `CompositeAdaptedChangeValidator` or something like that?

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @effiban 
I didn't do any change

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @effiban 
This name CompositeAdaptedChangeValidator does not say to the user why who need it

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @dimagorelik - so I think `CompositeChangeValidator` should be enough, and there should eventually only be one composite (I'm not sure but I think we can get rid of "compound" eventually

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @effiban 
I can explain why we need 2 such validators. Let's speak

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/OptimizedChangesValidator.java:18]: @effiban @galkoren 
One more name suggestion - TriggeredByChangesValidator

**Pull request closed without merging**